src/ripple/app/reporting/ETLSource.h

```diff
@@ -0,0 +1,435 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of rippled: https://github.com/ripple/rippled
+    Copyright (c) 2012, 2013 Ripple Labs Inc.
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#ifndef RIPPLE_CORE_ETLSOURCE_H_INCLUDED
+#define RIPPLE_CORE_ETLSOURCE_H_INCLUDED
+#include <ripple/app/main/Application.h>
+#include <ripple/app/reporting/ETLHelpers.h>
+#include <ripple/protocol/STLedgerEntry.h>
+#include <ripple/rpc/Context.h>
+
+#include <boost/algorithm/string.hpp>
+#include <boost/beast/core.hpp>
+#include <boost/beast/core/string.hpp>
+#include <boost/beast/websocket.hpp>
+
+#include "org/xrpl/rpc/v1/xrp_ledger.grpc.pb.h"
+#include <grpcpp/grpcpp.h>
+
+namespace ripple {
+
+class ReportingETL;
+
+/// This class manages a connection to a single ETL source. This is almost
+/// always a p2p node, but really could be another reporting node. This class
+/// subscribes to the ledgers and transactions_proposed streams of the
+/// associated p2p node, and keeps track of which ledgers the p2p node has. This
+/// class also has methods for extracting said ledgers. Lastly this class
+/// forwards transactions received on the transactions_proposed streams to any
+/// subscribers.
+class ETLSource
```

Comment:
I recommend making both `mutex` members `mutable` and then making every member function that you can `const`.  This gives the reader more information about which member functions change the state of the object.

Reply:
Changed the mutex members to `mutable` and added `const` to as many member functions as I could